mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala

```diff
@@ -0,0 +1,571 @@
+package scala.meta.internal.pc
+import org.eclipse.lsp4j.SemanticTokenModifiers
+import org.checkerframework.common.returnsreceiver.qual.This
+import scala.collection.mutable.ListBuffer
+import java.util
+import java.util.logging.Logger
+import java.{util => ju}
+import scala.meta.internal.jdk.CollectionConverters._
+import scala.meta.pc.VirtualFileParams
+import org.eclipse.lsp4j.SemanticTokenTypes
+import scala.meta.tokens._
+import scala.reflect.internal.util.SourceFile
+
+
+/**
+ * Corresponds to tests.SemanticHighlightLspSuite
+ */
+final class SemanticTokenProvider  (
+    protected val cp:MetalsGlobal // compiler
+  , val params: VirtualFileParams
+  , val capableTypes: util.List[String]
+  , val capableModifiers: util.List[String]
+)  {
+
+  // alias for long notation
+  def getTypeId(p:String):Int = capableTypes.indexOf(p)
+
+  // log tools
+  val logger = Logger.getLogger(classOf[This].getName)
+  val strSep = ", "
+  val linSep = "\n"
+
+  // initialize semantic tree
+  val ( unit,
+        root:cp.Tree, 
+        source:SourceFile,
+      )={
+        val unit = cp.addCompilationUnit(
+          params.text(),
+          params.uri().toString(),
+          None
+        )
+        cp.typeCheck(unit) // initializing unit
+
+        ( unit,
+          unit.lastBody,
+          unit.source
+        )
+      }
+
+  def unitPos(offset:Int)=unit.position(offset)
+  val nodes:Set[NodeInfo]=traverser.traverse(Set.empty[NodeInfo], root) 
+
+  
+
+
+  /** main method  */
+  def provide(): ju.List[Integer] =  {
+
+    // logger.info(linSep + linSep + params.text() + linSep)
+    // pprint.log(root)
+    //  Thread.sleep(3000)
+
+    var logString = linSep + params.text()
+    logger.info(treeDescriber(root) + linSep)
+
+    // logString += "\n\n nodes:" + nodes.size.toString()
+    // logString += nodes.toArray.sortBy(_.pos.start)
+    //               .map(n=>treeDescriber(n.tree,false))
+    //               .mkString("")
+
+    // Tools used later
+    val buffer = ListBuffer.empty[Integer]
+    var currentLine = 0
+    var lastLine = 0
+    var lastNewlineOffset = 0
+    var lastCharStartOffset = 0
+
+    def deltaLine():Int = currentLine - lastLine
+    def deltaStartChar(startPos:Int):Int={
+        if (deltaLine() == 0) startPos - lastCharStartOffset
+        else startPos - lastNewlineOffset
+    }
+    def addTokenToBuffer(
+      startPos:Int,charSize:Int,
+      tokenType:Int,tokeModifier:Int
+    ):Unit={
+      if (tokenType == -1 && tokeModifier == 0) return
+
+      buffer.addAll(
```

Comment:
if you replace all usages of `addAll` to `++=`, it will work on 2.11.12

Reply:
I reflect it.